src/main/scala/org/broadinstitute/dsde/workbench/avram/dataaccess/HttpSamDao.scala

```diff
@@ -0,0 +1,34 @@
+package org.broadinstitute.dsde.workbench.avram.dataaccess
+
+import java.util.logging.Logger
+
+import com.softwaremill.sttp._
+import io.circe.Decoder.Result
+import io.circe.{DecodingFailure, Json, ParsingFailure}
+import io.circe.generic.auto._
+import io.circe.parser._
+import org.broadinstitute.dsde.workbench.avram.util.{ErrorResponse, RestClient}
+
+class HttpSamDao extends SamDao with RestClient {
+  private val log: Logger = Logger.getLogger(getClass.getName)
+  // TODO: read from config
+  val samUrl = "https://sam.dsde-dev.broadinstitute.org"
+
+  override def getUserStatus(token: String): Either[ErrorResponse, UserInfo] = {
+    val request = buildAuthenticatedGetRequest(samUrl, "/register/user/v1?userDetailsOnly=true", token)
+    val response: Id[Response[String]] = request.send()
```

Comment:
I was super verbose here with types and intermediate vars just to keep everything straight. I definitely want to reduce the verbosity here. I think sttp has a way of plugging in serialization/deserialization so we don't have to do that explicitly. There may also be some way to consolidate the range of types used in the `Left`s so we can simply `map` instead of `match`.

Reply:
Yeah, definitely try to use `map`/`flatMap` instead of `match`. I think this could definitely be a for-comprehension:
```
val userInfoOrError = for {
  content <- response.body
  json <- parse(content)
  userInfo <- json.hcursor.get[UserInfo]("userInfo")
} yield userInfo

userInfoOrError.fold {
  error => // handle error
  result => result
}
```

Didn't compile but you get the idea..

Reply:
Yup. Just need to harmonize the error types, which is what I was illustrating by including the type annotations. They're all coming from different frameworks. sttp's `request.send()` gives a `String`, circe's `parse()` gives a `ParsingFailure`, and circe's `get()` gives a `DecodingFailure`. Circe's errors might be enough alike to map over (as long as they share the definition of `message()`), but I think it would be a problem to include sttp's `String` error in that for comprehension, right?

Reply:
Ohh right, the error types need to match. I wonder if we could define an `AvramException` type and automatically (implicitly?) convert the various cicre/sttp errors to the unified exception type.

Reply:
That sounds like a good idea and is the kind of feedback/advice I was hoping for. Thanks! I'm also absolutely thrilled that I understood what was going on well enough to recognize and explain it in a way that made sense! #fp_takeover_plot